### PR TITLE
Added getAt function to collections

### LINKED ISSUE #2138: Add Collection.getAt
# 🙋 Feature Request

Add a way to do index-based lookups to `@react-types` `Collection`.
 
`Collection`'s docs, "A generic interface to access a readonly sequential collection of unique keyed items", make it sound like doing an index-based lookup would be kosher.

## 🤔 Expected Behavior

Just do a lookup.

## 😯 Current Behavior

We have to convert `collection.getKeys` to an array, and do an index-based look up on the array to get the key, then `Collection.getItem(key)`:

```tsx
        <Virtuoso
          totalListHeightChanged={setListHeight}
          totalCount={state.collection.size}
          itemContent={(idx) => {
            // MapIterator doesn't have at/index lookup so make a copy
            const keys = [...state.collection.getKeys()];
            const item = state.collection.getItem(keys[idx]);
            if (item) {
              return <Option key={item.key} item={item} state={state} />;
            }
          }}
        />
```

## 🔦 Context

We're integrating `useListBox` / `useMultipleSelectionState` with react-virtuoso for virtualized rendering, and the virtuoso api has a `itemContent: (index) => ReactNode` method, that we want to implement by doing `state.collection.getAt(index)`, where `state` is a `SelectState`.

### PR DESCRIPTION
**Pull request by lukerhoads**
Closes #2138

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

Seemed like too easy of an issue to pass up

### CODE REVIEW

**Comment by dannify** [comment]
@lukerhoads Can you please sign the CLA? You will need to close and open the PR again to trigger it.

File: packages/@react-stately/grid/src/GridCollection.ts [hunk 1]
```diff
@@ -150,4 +150,9 @@ export class GridCollection<T> implements IGridCollection<T> {
   getItem(key: Key) {
     return this.keyMap.get(key);
   }
+
+  getAt(idx: number) {
```
Comment (devongovett) [inline packages/@react-stately/grid/src/GridCollection.ts:154]: Can we call this just `at` to match the new [Array#at](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at) function being added to JavaScript?

File: packages/@react-types/shared/src/collections.d.ts [hunk 2]
```diff
@@ -130,6 +130,9 @@ export interface Collection<T> extends Iterable<T> {
   /** Get an item by its key. */
   getItem(key: Key): T,
 
+  /** Get an item by the index of its key. */
+  getAt(idx: number): T,
```
Comment (dannify) [inline packages/@react-types/shared/src/collections.d.ts:134]: Missed the rename of this one?

Comment (lukerhoads) [reply packages/@react-types/shared/src/collections.d.ts]: My bad. Just fixed. Next time if this happens you guys can open a PR. This is prob why open sourcing is annoying to the more experienced ppl 😂

Comment (dannify) [reply packages/@react-types/shared/src/collections.d.ts]: Not a problem, we are here to help! Thanks for the PR

**Review by dannify — APPROVED** [review]

**Review by devongovett — APPROVED** [review]

**Pull request merged**